### PR TITLE
Use TextIOWrapper.detach() instead of NoCloseStream

### PR DESCRIPTION
**Pull request by jdufresne**
Usage and this pattern is discussed in Python bug: https://bugs.python.org/issue21363

[detach() docs](https://docs.python.org/3/library/io.html#io.BufferedIOBase.detach)

### CODE REVIEW

**Comment by jdufresne** [comment]
Rebased on latest mater.
